packages/expo-permissions/src/PermissionsHooks.ts

```diff
@@ -0,0 +1,48 @@
+import { useCallback, useEffect, useState } from 'react';
+
+import { PermissionType, PermissionResponse, askAsync, getAsync } from './Permissions';
+
+/**
+ * Get or ask permission for protected functionality within the app.
+ * It returns the permission response after fetching or asking it.
+ * The hook fetches the permissions when rendered, by default.
+ * To ask the user permission, use the `askPermission` callback or `ask` option.
+ *
+ * @see https://docs.expo.io/versions/latest/sdk/permissions/
+ * @example const [permission, askPermission, getPermission] = usePermissions(...);
+ */
+export function usePermissions(
+  type: PermissionType | PermissionType[],
+  options: PermissionsOptions = {}
+): [PermissionResponse | undefined, () => Promise<void>, () => Promise<void>] {
+  const [data, setData] = useState<PermissionResponse>();
+  const { ask = false, get = true } = options;
+  const types = Array.isArray(type) ? type : [type];
+
+  // note: its intentional to listen to `type`, not `types`.
+  // when `type` is casted to an array, it possible creates a new one on every render.
+  // to prevent unnecessary function instances we need to listen to the "raw" value.
+
+  const askPermissions = useCallback(() => askAsync(...types).then(setData), [type]); // eslint-disable-line react-hooks/exhaustive-deps
+
+  const getPermissions = useCallback(() => getAsync(...types).then(setData), [type]); // eslint-disable-line react-hooks/exhaustive-deps
```

Comment:
What would happen here if those `askAsync` or `getAsync` will reject? 🤔

Reply:
Yeah, that's still on my todo list. I'm not 100% how we could handle errors like these. Generally speaking, these calls should never reject but you never know what would happen 😅  

Currently, they would classify as unhandled global promise rejections. At least, they would show up in Sentry, but it's not a "perfect" way of developing I guess. If we want to keep track of those errors, we could add them to a state and give the dev access to it.

Reply:
Also, to clarify why `types` isn't used here in the `useCallback` dependencies:

- I cast the provided types to an array. In case you provide only 1 type, it's still casted to array to simplify calling the permissions api.
- Because I cast it to an array, this value could be a new array every update. This would falsy assume these callbacks should be updated.

That's why I check on the raw input value instead 😄

Reply:
`askAsync` or `getAsync` can reject in two different situations:
- when you don't have the required requester (on the iOS). 
- if you asked about permission and the app doesn't have active activity (on the Android).
The second one, we can ignore. However, we should handle rejection in some way because of the first case.

Reply:
@lukmccall Thanks, so if I understand properly, these methods may fail when native stuff isn't available. How would a user mitigate this? I think this is important to implement proper error handling. If the user has to restart the app, I would prefer an `onError` callback to the options. This way the user can execute code immediately, instead of checking on the `error` object returned. But if there is no mitigation possible, returning the error object for remote logging might be enough.

Reply:
These methods may fail if the user didn't have proper requesters which are located in a different package. For example, the requester for the camera permission is in the `expo-camera`. So the developer needs to install this module to ask/get this permission - the `onError` callback sound good to me in this case 👍